sharktank/sharktank/ops/qconv_impls.py

```diff
@@ -0,0 +1,230 @@
+                # add directly in integer. A quantized bias cannot be arbitrary and must
+                # be a symmetric quantization of the output scale. This is not verified
+                # and driven by the data.
+                # Broadcast the bias scale to the channels in the NCHW output.
+                rescale_d = bias_layout.d.reshape(1, -1, 1, 1)
+            else:
+                warnings.warn(f"unsupported qconv bias quantization: {bias_layout}")
+
+    # Alias components (d=scale, qs=quantized samples, m=offset).
+    if accum_dtype is None:
+        accum_dtype = torch.int32
+    input_d = input_layout.d
+    input_dtype = input_layout.dtype
+    input_qs = input_layout.qs
+    input_m = input_layout.m
+    weight_d = weight_layout.d
+    weight_qs = weight_layout.qs
+    weight_m = weight_layout.m
+
+    # Verify that quantization axis meets our requirements.
+    flat_input_d, flat_input_m = _flatten_input_scale_offset_channels(input_d, input_m)
+    flat_weight_d, flat_weight_m = _flatten_weight_scale_offset_channels(
+        weight_d, weight_m
+    )
+    if flat_input_d is None or flat_weight_d is None:
+        return NotImplemented
+
+    # No quantized bias: infer the output rescale.
+    if rescale_d is None:
+        # Output scale by the product of input and weight scale and broadcast
+        # to the output NCHW shape.
+        rescale_d = (input_d * flat_weight_d).reshape(1, -1, 1, 1)
+
+    # TODO: Use a real mixed precision op.
+    y_qs = torch.nn.functional.conv2d(
+        input_qs.to(dtype=torch.float32),
+        weight_qs.to(dtype=torch.float32),
+        bias=bias_qs.to(dtype=torch.float32),
+        stride=stride,
+        padding=padding,
+        dilation=dilation,
+        groups=groups,
+    ).to(dtype=accum_dtype)
+
+    # Apply offset corrections.
+    if input_m is not None:
+        # Apply offset correction for asymmetric input.
+        # At the time of this writing, this was not a common case.
+        input_offset_fix = torch.sum(
+            torch.flatten(weight_qs, 1), dim=1, dtype=accum_dtype
+        ).unsqueeze(0) * flat_input_m.unsqueeze(1)
+        input_offset_fix = input_offset_fix.unsqueeze(2).unsqueeze(3)
+        y_qs = y_qs - input_offset_fix
+    if flat_weight_m is not None:
+        # Apply offset correction for asymmetric weight.
+        # At the time of this writing, this was the common case.
+        # The weight_m shape is the offset relative to weight and needs to
+        # be reshaped to broadcast to the NCHW output. Since all but the
+        # output channels are zero, just
+        weight_offset_fix = torch.nn.functional.avg_pool2d(
+            input_qs.to(dtype=torch.float32),
+            (weight_qs.shape[2], weight_qs.shape[3]),
+            stride=stride,
+            padding=padding,
+            divisor_override=1,
+        ).to(dtype=accum_dtype)
+        weight_offset_fix = weight_offset_fix.sum(1, keepdim=True, dtype=accum_dtype)
```

Comment:
Consider moving this summation to be on `input_qs` along the same dimension (with keeps dims). The operation should be separable and should improve the conv performance.